borg/crypto.pyx

```diff
@@ -63,93 +74,222 @@ def num_aes_blocks(int length):
-        # This is what the openssl docs say. I am not sure this is correct,
-        # but OTOH it will not cause any harm if our buffer is a little bigger.
-        cdef unsigned char *out = <unsigned char *>malloc(inl+16)
-        if not out:
+        cdef Py_buffer idata = ro_buffer(data)
+        cdef int ilen = len(data)
+        cdef unsigned char *odata = <unsigned char *>malloc(ilen + 16)
+        cdef int olen
+        cdef int offset
+        if not odata:
             raise MemoryError
         try:
-            if not EVP_DecryptUpdate(&self.ctx, out, &outl, <const unsigned char*> data_buf.buf, inl):
+            if not EVP_DecryptInit_ex(&self.ctx, EVP_aes_256_gcm(), NULL, NULL, NULL):
+                raise Exception('EVP_DecryptInit_ex failed')
+            iv = self.fetch_iv(<unsigned char *> idata.buf)
+            # Set IV length (bytes)
+            if not EVP_CIPHER_CTX_ctrl(&self.ctx, EVP_CTRL_GCM_SET_IVLEN, 16, NULL):
+                raise Exception('EVP_CIPHER_CTX_ctrl SET IVLEN failed')
+            if not EVP_DecryptInit_ex(&self.ctx, NULL, NULL, self.enc_key, iv):
+                raise Exception('EVP_DecryptInit_ex failed')
+            # Set expected tag (mac) value. Works in OpenSSL 1.0.1d and later.  <-- XXX WAT!?
+            if not EVP_CIPHER_CTX_ctrl(&self.ctx, EVP_CTRL_GCM_SET_TAG, 16, <void *> idata.buf):
+                raise Exception('EVP_CIPHER_CTX_ctrl SET TAG failed')
+            if not EVP_DecryptUpdate(&self.ctx, NULL, &olen, <const unsigned char*> idata.buf+16, 16):
+                raise Exception('EVP_DecryptUpdate failed')
+            offset = 0
+            rc = EVP_DecryptUpdate(&self.ctx, odata+offset, &olen, <const unsigned char*> idata.buf+32, ilen-32)
+            if not rc:
                 raise Exception('EVP_DecryptUpdate failed')
-            ptl = outl
-            if EVP_DecryptFinal_ex(&self.ctx, out+ptl, &outl) <= 0:
+            offset += olen
+            rc = EVP_DecryptFinal_ex(&self.ctx, odata+offset, &olen)
+            if rc <= 0:
                 # this error check is very important for modes with padding or
                 # authentication. for them, a failure here means corrupted data.
-                # CTR mode does not use padding nor authentication.
-                raise Exception('EVP_DecryptFinal failed')
-            ptl += outl
-            return out[:ptl]
+                # for GCM mode, a failure here means corrupted / tampered tag (mac) or data
+                raise Exception('EVP_DecryptFinal_ex failed')
+            offset += olen
+            return odata[:offset]
         finally:
-            free(out)
-            PyBuffer_Release(&data_buf)
+            free(odata)
+            PyBuffer_Release(&idata)
+
+    cdef fetch_iv(self, unsigned char * idata):
+        return idata[16:32]
+
+    cdef store_iv(self, unsigned char * odata, unsigned char * iv):
+        cdef int i
+        for i in range(16):  # XXX easier way?
```

Comment:
Cython bug?

```py
    cdef store_iv(self, unsigned char * odata, unsigned char * iv):
        odata[16:32] = iv
```

=

```
  /* "borg/crypto.pyx":290
 * 
 *     cdef store_iv(self, unsigned char * odata, unsigned char * iv):
 *         odata[16:32] = iv             # <<<<<<<<<<<<<<
 *         #cdef int i
 *         #for i in range(16):  # XXX easier way?
 */
  __pyx_t_1 = __Pyx_PyBytes_FromCString(__pyx_v_iv); if (unlikely(!__pyx_t_1)) __PYX_ERR(0, 290, __pyx_L1_error)
  __Pyx_GOTREF(__pyx_t_1);
  if (__Pyx_PyObject_SetSlice(((PyObject *)__pyx_v_odata), __pyx_t_1, 16, 32, NULL, NULL, NULL, 1, 1, 1) < 0) __PYX_ERR(0, 290, __pyx_L1_error)
  __Pyx_DECREF(__pyx_t_1); __pyx_t_1 = 0;
```

__pyx_v_odata is the unsigned char* as passed in.

Wtf?